euphonic/design/03_design_document_v0.3.md

```diff
@@ -0,0 +1,531 @@
+# Euphonic v0.3 Design Document
+
+# Requirements and Context
+Euphonic is a piece of software developed under PACE to simulate inelastic
+neutron scattering data for phonons calculated from a force constants matrix.
+There are 3 main use cases (+1 that should be considered), each with their own
+requirements:
+
+## 1. Horace/Brille
+
+Horace has several simulation functions that require a handle to a user function
+which has certain defined inputs and outputs. `disp2sqw` requires a user
+function that should take a list of 3d q-points as input and return a list of
+structure factors and frequencies for each phonon mode and q-point, which
+Euphonic should be able to provide. Euphonic uses Fourier interpolation to
+calculate phonons at every q-point which is accurate but expensive, so another
+piece of software is being developed, Brille, which will use Euphonic to create
+a lookup table of phonon eigenvectors and frequencies, which can be used to
+cheaply get frequencies and eigenvectors on a finer grid via linear
+interpolation. The structure factor can then be calculated on this finer grid
+(calculating the structure factor is much cheaper than calculating the
+frequencies/eigenvectors). The hierarchy of Horace/Brille/Euphonic is not yet
+decided so Euphonic must be flexible. In addition, in the future there is scope
+to calculate the structure factor for powder samples, this also needs to be
+considered. From this the requirements are:
+
+- Euphonic must be able to take a list of q-points as input, and return the
+crystal structure factor, and phonon frequencies for each q-point
```

Comment:
It's not clear from this description how Euphonic should get the necessary information about the lattice dynamics. Does Horace 
- also specify the path to a phonon data file,
- pass a previously instantiated object containing lattice parameters and force constants or
- require Euphonic to hold an internal state that has previously been set to the system of interest?

Reply:
This part is mainly describing what the required inputs/outputs are for a Horace simulation function. You must pass to Horace a function which takes qh, qk, ql at a minimum (plus whatever parameters you like) and returns the structure factor, and mode frequencies at those q-points. The extra parameters to that function and what happens inside it are sort of implementation details and don't matter to Horace.

In reality we'd probably get a Horace user to input a model type (e.g. 'CASTEP' or 'Phonopy') then allow whatever keyword arguments apply to the from_castep, from_phonopy methods, where Euphonic would then read and interpolate the force constants. But that could be changed depending on if Brille is an intermediate, we end up needing to write to file etc... As long as Euphonic can take q-points and return structure factors and mode frequencies that's all Horace requires.

Reply:
I think we'd have a function on the Horace side that makes all the calls to Euphonic required to go from qpts -> frequencies + structure factors. I forgot to mention there's an example of that function [here](https://github.com/pace-neutrons/Meuphonic/blob/master/matlab/euphonic/euphonic_sf.m). The Python/Matlab interface there is terrible and I would definitely rewrite it, but you get the idea.